File: J9Ch/src/J_8_Data_Structure/Optional_6/_471_Top_K_Frequent_Words_Medium.py
# /** 471 Top K Frequent Words


class Solution:
    # @param {string[]} words a list of string
    # @param {int} k an integer
    # @return {string[]} a list of string
    def topKFrequentWords(self, words, k):
        # Write your code here
        dict = {}
        for word in words:
            if word not in dict:
                dict[word] = 1
            else:
                dict[word] += 1
        p = []
        for key, value in dict.items():
            p.append((value, key))

        p.sort(cmp=self.cmp)
        result = []
        for i in xrange(k):
            result.append(p[i][1])
        return result

    def cmp(self, a, b):
        if a[0] > b[0] or a[0] == b[0] and a[1] < b[1]:
            return -1
        elif a[0] == b[0] and a[1] == b[1]:
            return 0
        else:
            return 1

File: J9Ch/src/J_8_Data_Structure/Optional_6/python_cpp/_471_Top_K_Frequent_Words_Medium.cpp
/** 471 Top K Frequent Words */


class Node {
public:
    int time;
    string word;
    Node(int _time, string _word) {
        time = _time, word = _word;
    };
    bool operator<(const Node& obj) const {
        return time > obj.time || time == obj.time && word < obj.word;
    }
};
class Solution {
public:
    /**
     * @param words an array of string
     * @param k an integer
     * @return an array of string
     */
    vector<string> topKFrequentWords(vector<string>& words, int k) {
        // Write your code here
        int n = words.size();
        map<string, int> mp1;
        map<string, int>::iterator it1;
        for (int i = 0; i < n; ++i) {
            if (mp1.find(words[i]) == mp1.end())
                mp1[words[i]] = 1;
            else
                mp1[words[i]] += 1;
        }
        vector<Node> p;
        for (it1 = mp1.begin(); it1 != mp1.end(); it1++)
            p.push_back(Node(it1->second, it1->first));
        sort(p.begin(), p.end());
        vector<string> result;
        for (int i = 0; i < k; ++i)
            result.push_back(p[i].word);
        return result;
    }
};